Added Dis Max Query

mszyndel opened the pull request:
As my company really needs dis max I added it.
Please let me know if I should write some more tests; units should be sufficient, but there are just two basic integration tests.

Commit:
added dis_max query

Commit:
added dis_max query

Commit:
added dis_max query

erickt committed:
[#352] Whitespace cleanup in Tire::Results::Collection

erickt committed:
[#352] Added MyModel#new_record? for Tire::Persistence

Commit <https://github.com/erickt/tire/commit/40cb9753b25cd943e35ddf5ce52c0384ae98482e> rewritten and cleaned up by @karmi

erickt committed:
[#352] Added, that accessing Tire::Search#json forces perform

Conflicts:

	lib/tire/search.rb

Commit:
Merge branch 'master' of git://github.com/karmi/tire into upstream/master

ahfeel committed:
[#366] Added support for the Update API in Tire::Index

Usage:

    tags = ['a', 'b', 'c']

    Tire.index 'my_index' do
      update 'document_type', 'document_id', {
        :script => "_ctx._source.tags = tags;',
        :params => {:tags => tags}
      }
    end

erickt commented:
Hi hajder! I think tire HEAD recently grew support for aliases. Any chance you could rebase onto head and update your patch?

andruby committed:
Added, that boost factor can be specified for indices

Example:

    Tire.search 'index_1' => { boost: 1.5 }, 'index_2' => { boost: 2 } do
      ...
    end

Closes #205, closes #377.

mszyndel committed:
Merge remote-tracking branch 'upstream/master'

mszyndel commented:
@erickt done.

Commit:
Added that MyModel#all is called in Rake import task

Closes #380, closes #381.

karmi committed:
[#366] Cleaned up the Update API support and added an integration test suite

This patch expands karmi/tire@f620b36.

* Polished the `curl` output from the API call
* Added the ability to pass URL parameters (timeout, refresh, etc)
* Cleaned up the unit tests
* Added a separate integration test for resilience and documentation

Example usage
-------------

Update a counter:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ruby
  Tire.index('articles-with-tags') { update( 'article', '1', :script => "ctx._source.views += 1" ) and refresh }
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Add a tag to the document:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ruby
    Tire.index('articles-with-tags') do
      update 'article', '1', :script => "ctx._source.tags += tag",
                             :params => { :tag => 'new-tag' }
      refresh
    end
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Set a title of the document:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ruby
  @title = 'Foo Is Now Bar!'

  Tire.index('articles-with-tags') do |index|

    index.update 'article', '1', :script => "ctx._source.title = title",
                                 :params => { :title => @title }
    index.refresh
  end
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

See <http://www.elasticsearch.org/guide/reference/api/update.html> for more documentation.

---

Closes #366.

karmi committed:
Patch Test::Unit::Assertion#assert on Ruby 1.8 to return true on success

Our Mocha `with do ... end` blocks fail otherwise.

karmi committed:
[TESTS] Fixed errors on Ruby 1.8

* Changed Ruby 1.9 hashes to hashrocket syntax
* Changed regex to allow different URL parts order

karmi committed:
[CI] Added running functional tests against live elasticsearch at Travis CI [http://travis-ci.org/#!/karmi/tire]

Squashed from the `karmi/tire@travis` branch:

c2d7f0d ‣ Restructuring the Travis configuration
79f454d ‣ Added running integration tests against live elasticsearch on Travis
d1c4374 ‣ Allow build failures on 1.8.7 and REE
b4fa8a0 ‣ Changed, that integration tests are run only on Ruby 1.9, unit tests on all Rubies

karmi committed:
[CI] Use "turn" when running under CI

mszyndel commented:
Yes, order of results is guaranteed.
I guess it would be nice if you read code before commenting on pull requests.

karmi commented:
@cwalsh You mean in tests? Yeah, the order should absolutely be guaranteed.

@hajder Don't let your bitterness get in front of polite discussion :) @cwalsh was just asking a question .)

karmi commented on test/integration/dis_max_queries_test.rb:
```diff
@@ -0,0 +1,43 @@
+  class DisMaxQueriesIntegrationTest < Test::Unit::TestCase
+    include Test::Integration
+
+    context "Boolean queries" do
```
Should be `Dis Max queries`.

karmi commented:
I guess an integration test in this case would be great:
- It would clear any confusion (eg. about the guaranteed ordering, see discussion above)
- Integration tests serve as a great documentation on the exact purpose of the feature, not only as a "defensive mechanism". See eg. https://github.com/karmi/tire/blob/master/test/integration/index_update_document_test.rb#L25-35 [karmi/tire@ff00015] for an example for the _Update API_ which covers the usual use-cases for the API.

karmi committed:
[FIX] Fixed incorrect `Results::Item#to_hash` serialization

Without this fix, `Item#to_hash` returned `Item` instances for nested Hashes.

This commit calls `to_hash` recursively in `Item#to_hash`

karmi committed:
Fixed incorrect `to_curl` output in Tire::Index#register_percolator_query

karmi committed:
[#190] [FIX] Tire::Search::Search#version should return `self` to allow chaining

Fixes karmi/tire@2bf558508e0cf8457c826e7251e80a501ed0340e by @gedwards

karmi committed:
[ACTIVEMODEL] Allow to pass :version option in Tire::Model::Search::ClassMethods#search

karmi committed:
[#352] [FIX] Fixes MyModel#persisted? for Tire::Model::Persistence integration

Fixes karmi/tire@cef81338e225150dea0181773760647f05cf83e0 by @erickt

karmi committed:
[FIX] Fixes Ruby 1.8.7 incompatibility

* Fixes karmi/tire@6c3b852 by @karmi (incorrect trailing dot)
* Fixes karmi/tire@ce47d00 by @karmi (Hash order)

karmi commented:
Additional info for the integration test -- there's a great example of "albino elephant" at the http://www.elasticsearch.org/guide/reference/query-dsl/dis-max-query.html page, which should be easy to translate to a _Tire_ integration test:

> If the query is “albino elephant” this ensures that “albino” matching one field and “elephant” matching another gets a higher score than “albino” matching both fields.

Commit:
Fix typo (s/operation/operating/) system

Closes #412

mszyndel committed:
Merge branch 'master' of https://github.com/hajder/tire

mszyndel committed:
Merge branch 'master' of https://github.com/hajder/tire

mszyndel committed:
improved tests

mszyndel commented:
I wrote an additional test to check order of results returned, however it's not an "albino elephant" example (I belive this would be too dependent on elasticsearch internals).

@karmi I'm not completely sure if I did rebase right, so please take a look :)

karmi commented:
@hajder Thanks, merged.

@hajder, @nbudin Guys, maybe you can shed some light on your `dis_max` usage for me? As you can see from note in karmi/tire@7431ce9, the `dis_max` query gives the same result as a plain old `boolean` query? Can you share some example where it does make a difference from your application/domain? Thanks!

nbudin commented:
Interesting.  I hadn't been aware that a `boolean` query with all `should` clauses and `minimum_number_should_match: 1` would give the same result as a `dis_max` query; I had assumed the same thing you had at the bottom of that comment: that `boolean` would sum the scores.

I do think `dis_max` has the advantage of being both more specific and simpler.  It's also less likely to change in future ElasticSearch updates - not that `boolean` is necessarily likely to change, but `dis_max` should always mean dismax.
